substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrThreadLocalMemory.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2021, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.jfr;
+
+import org.graalvm.word.WordFactory;
+
+import com.oracle.svm.core.annotate.Uninterruptible;
+import com.oracle.svm.core.locks.VMMutex;
+
+public class JfrThreadLocalMemory {
+    private static JfrBuffer head;
+    private static final VMMutex mutex = new VMMutex();
+
+    @Uninterruptible(reason = "Accesses a JFR buffer.")
+    public static JfrBuffer acquireThreadLocalBuffer(long bufferSize) {
+        JfrBuffer node = JfrBufferAccess.allocate(WordFactory.unsigned(bufferSize));
+        node.setNext(head);
+        head = node;
+        return node;
+    }
+
+    @Uninterruptible(reason = "Accesses a JFR buffer.")
+    public static boolean removeThreadLocalBuffer(JfrBuffer buffer) {
+        mutex.lockNoTransition();
+        try {
+            if (buffer.isNull()) {
+                return false;
+            }
+
+            if (buffer.equal(head)) {
+                head = head.getNext();
+                JfrBufferAccess.free(buffer);
+                return true;
+            }
+            JfrBuffer prev = head;
+            JfrBuffer node = head.getNext();
+            while (node.isNonNull()) {
+                if (buffer.equal(node)) {
+                    prev.setNext(node.getNext());
+                    JfrBufferAccess.free(buffer);
+                    return true;
+                }
+                prev = node;
+                node = node.getNext();
+            }
+
+            return false;
+        } finally {
+            mutex.unlock();
+        }
+    }
+
+    public static void writeThreadLocalBuffers(JfrChunkWriter writer) {
+        mutex.lock();
```

Comment:
This lock may cause deadlocks:
- thread A executes `writeThreadLocalBuffers` - locks the mutex
- thread B executes `acquireThreadLocalBuffer` - blocks in uninterruptible code, when trying to lock the mutex
- thread A triggers a safepoint in `writeThreadLocalBuffers` - a deadlock occurs as the safepoint can't be initiated because thread B is still in uninterruptible code.

Reply:
Hmm, sounds like I really do want a non-blocking linked list. Maybe something like https://www.cl.cam.ac.uk/research/srg/netos/papers/2001-caslists.pdf

Reply:
Or be uninterruptible when writing the thread local buffers...

Reply:
It would also be sufficient to ensure that you only hold the lock while in uninterruptible code, i.e., you could unlock the lock after acquiring a buffer that contains data that needs to be persisted. After unlocking the mutex, you can call interruptible code.

Reply:
Okay, good point. I'm working through this now and will @ when ready. I need to think this through more; currently the critical section is the entire list as iteration needs to account for `removeBuffer` calls when threads die.